PG.StarWarsGame.Files.MEG/PG.StarWarsGame.Files.MEG/Services/V1/MegFileProcessService.cs

```diff
@@ -0,0 +1,251 @@
+    ///     implementing <see cref="Microsoft.Extensions.Logging.ILoggerFactory" />
+    /// </summary>
+    [Export(nameof(IMegFileProcessService))]
+    public sealed class MegFileProcessService : IMegFileProcessService
+    {
+        [NotNull] private readonly IFileSystem m_fileSystem;
+        [CanBeNull] private readonly ILogger m_logger;
+
+        public MegFileProcessService([CanBeNull] IFileSystem fileSystem,
+            [CanBeNull] ILoggerFactory loggerFactory = null)
+        {
+            m_fileSystem = fileSystem ?? new FileSystem();
+            m_logger = loggerFactory?.CreateLogger<MegFileProcessService>();
+        }
+
+        public void PackFilesAsMegArchive(string megArchiveName,
+            Dictionary<string, string> packedFileNameToAbsoluteFilePathsMap, string targetDirectory)
+        {
+            if (!StringUtility.HasText(megArchiveName))
+            {
+                throw new ArgumentException(
+                    "The provided argument is null, an empty string or only consists of whitespace.",
+                    nameof(megArchiveName));
+            }
+
+            if (!packedFileNameToAbsoluteFilePathsMap.Any())
+            {
+                throw new ArgumentException(
+                    "The provided argument does not contain any elements.",
+                    nameof(packedFileNameToAbsoluteFilePathsMap));
+            }
+
+            if (!StringUtility.HasText(targetDirectory))
+            {
+                throw new ArgumentException(
+                    "The provided argument is null, an empty string or only consists of whitespace.",
+                    nameof(targetDirectory));
+            }
+
+            string actualName = StringUtility.StripFileExtension(megArchiveName);
+            MegFileHolder megFileHolder = new MegFileHolder(targetDirectory, actualName);
+            foreach ((string key, string value) in packedFileNameToAbsoluteFilePathsMap)
+            {
+                megFileHolder.Content.Add(new MegFileDataEntry(key, value));
+            }
+
+            MegFileBuilder builder = new MegFileBuilder(m_fileSystem);
+            MegFile megFile = builder.FromHolder(megFileHolder, out List<string> filesToStream);
+            string writePath = m_fileSystem.Path.Combine(megFileHolder.FilePath, megFileHolder.FullyQualifiedName);
+            CreateTargetDirectoryIfNotExists(megFileHolder.FilePath);
+            using (BinaryWriter writer = new BinaryWriter(m_fileSystem.FileStream.Create(writePath, FileMode.Create)))
+            {
+                writer.Write(megFile.ToBytes());
+            }
+
+            foreach (string file in filesToStream)
+            {
+                using Stream readStream = m_fileSystem.File.OpenRead(file);
+                using Stream writeStream =
+                    m_fileSystem.FileStream.Create(writePath, FileMode.Append, FileAccess.Write, FileShare.None);
+                byte[] buffer = new byte[1024];
```

Comment:
i think the default buffer size in the .net implementations is 4k or 8k

Reply:
The default according to Microsoft for Stream to Stream seems to be 8192 bytes, but in any IO Stream it's 4096 bytes, so I'll go with that.